Add support for Lyrics

X-Ryl669 opened the pull request:
This is a large PR, and probably need some guidance to fix my errors.

This adds support for capturing lyrics and displaying them in the web interface, to play in sync (or not) with a song being played.
It appears like this:
![image](https://github.com/owntone/owntone-server/assets/3277165/8401cb98-830a-43b3-8d55-0533a53e609b)

The PR is split in 3 commits that are completely independent of each other.

### First commit
The first commit adds a key in the file table in the database to store the lyrics.
I'm using ffmpeg's metadata API here, so the lyrics are available in keys `lyrics_eng` or `lyrics_XXX`, etc (the key name contains the language for the lyric). This works with audio files with embedded LRC lyrics and non synchronized lyrics.

Since the current implementation used to look for fixed keys' name, I've inverted the key searching logic so that the code enumerate all metadata keys from the file and match them with the metadata key Owntone is able to handle. This avoid doing 2 pass over the metadata (one for usual metadata from Owntone's keys), and another for searching any lyrics.
My algorithm is doing a single O(N) search *like before*, so it's not changing the speed of Owntone's file parsing.

If it proves too slow anyway, we could force the Owntone's metadata keys array to be sorted by name and perform a binary search on them. 

I've tested all the parsing and database code.
Please check the db upgrade code, I'm not 100% sure it's correct and I wasn't able to test it correctly, since I don't know why it doesn't upgrade.

### Second commit
I think I got this right. This adds a `/api/library/lyrics` route to fetch a track's lyrics. The `api/library/track` route was modified to add a `"lyrics": 1` if lyrics are available for a track, with the idea to optimize a call to `/api/library/lyrics` if they aren't. The web interface doesn't use this route anyway, so it would only be useful for any other client.

I've tested the API completely

### Third commit
This adds the required changes to the web interface to display lyrics and add a lyrics pane.
I've added 2 components (`Lyrics.vue` and `PlayerButtonLyrics.vue`). The latter allow to toggle the visibility of the former.
I've modified the store to support lyrics too and the main application to synchronize fetching lyrics when the player status is updated. This means that an additional request is made each time a new track is started (to fetch the lyrics).

I had to modify the bottom navbar code too to deal with the lyrics toggling button. I've removed the hack used to center the play control icons (with auto margin left/margin right) and instead used the flex properties and made 3 groups (the left group contains the queue button, the center group contains the player control icons, and the right group contains the lyric toggle and output selection button).

What I'm not sure of:
1. In the `Lyrics.vue`, I've embedded a lot of style code that's required for animations and displaying the pane. The other components don't do this, I'm not sure what is the policy here or if I need to move it elsewhere.
2. Currently, the color is hardcoded in this embedded stylesheet. So any dark mode switching would require some tweaking. I don't know how you intend to deal with this.

If required, I can send you songs with the 2 kinds of embedded lyrics if you want to test.

ejurgensen commented:
Thanks for the PR. A general question from me would be if it is normal to have the lyrics part of the embedded metadata? If you adding those tags yourself from some online source, might it then be an idea to get the lyrics directly from there?

Anyway, as you say, the PR is large, so it might take some time to review it. I suggest you start by reviewing it yourself - check that code style is consistent with the rest of OwnTone, typos stuff like that. Check that MODULE_RAUTH also.

Afterwards, I think both @hacketiwack and I will need to review.

hacketiwack commented:
Yes, indeed, I need a bit of time to check that PR.

To answer your question @ejurgensen it can/may be normal to have lyrics embedded into the metadata.
At least, there is an extension for MusicBrainz Picard to do it and there are "official" tags for it.
The problem here is more: what is a reliable source of lyrics as there are not that many freely available nowadays.

X-Ryl669 commented:
> Is it normal to have the lyrics part of the embedded metadata? If you adding those tags yourself from some online source, might it then be an idea to get the lyrics directly from there?

There's 2 kind of lyrics: synchronized one and non synchronized lyrics. The former is usually using the LRC format. You can use [LRCGet](https://github.com/tranxuanthang/lrcget). The latter was easier to find and you could run [beets](https://github.com/beetbox/beets) that'd fetch them for you.

To embed any kind of lyrics you can use [mutagen library](https://mutagen.readthedocs.io/en/latest/) via a CLI tool like [eyeD3](https://eyed3.readthedocs.io/en/latest/) or a GUI tool like [Ex Falso](https://flathub.org/apps/io.github.quodlibet.ExFalso).

I haven't added support for checking if a .LRC file exists as a sibling of a .MP3 file in a folder. Some player do that, but it's much easier to embed them inside a single file instead.

X-Ryl669 commented:
> The problem here is more: what is a reliable source of lyrics as there are not that many freely available nowadays.

There's LyricsWiki where you'll find almost all (unsynchronized) lyrics (they have more than 300 000 lyrics in their database). When using beets, you can set many source to fetch lyrics (google, lyricswiki, musicmatch, genius, etc...)
With LRCGet, I don't know what's the source used, but it does find a lot of them (but not all).

X-Ryl669 commented:
@ejurgensen I've a lot of issue with the indentation rules used in Owntone. It's a mix of tab and spaces and no matter what I set in my editor, it never feel right. What are the indentation rules and, better, do you have a tool to indent with this rules ?
I've tried to follow the coding style as much as possible already (for C files I meant).

X-Ryl669 commented:
I've written this python script that follow this indentation rules:
```python
#!/bin/python
import sys


def replaceTab(line):
    spaceLine = line.rstrip().replace('\t', '        ')
    meat = spaceLine.lstrip()
    # Then convert left spaces to tab following the rules:
    indentation = len(spaceLine) - len(meat)
    tabCount = int(indentation / 8)
    return '\t' * tabCount + ' ' * int(indentation % 8) + meat

def main():
    dryRun = False
    if len(sys.argv) < 2:
        print("Usage is {} [-n] fileToFix".format(sys.argv[0]))
        return 0

    file = sys.argv[1]
    if (file == '-n'):
        dryRun = True
        file = sys.argv[2]

    print("Processing: {} {}".format(file, "dry run" if dryRun else ""))
    outLines = []
    with open(file, 'r') as f:
        lines = f.read().split('\n')
        for line in lines:
            outLines.append(replaceTab(line))

    if dryRun:
        for line in outLines:
            print(line)
    else:
        with open(file, 'w') as f:
            f.write('\n'.join(outLines))
    
if __name__ == '__main__':
    main()
```
And applied to `main.c`. There are multiple place in the source that doesn't follow this rule (off by 1 space or 2).
I haven't run this indentation tool on all files, it's probably out of sync in multiple places.

It's probably another PR request, but would you accept to switch to a more conventional indentation rule (like all spaces, 2 space per level) ?
I've written a clang-format config file but obviously, this kind of mixed indentation doesn't exist. And the *indent opening brace on their own line* requires the latest clang version.

Anyway, I've removed the RAUTH line, I don't remember putting it here. I think I've followed the other coding convention I'm seeing around. 
I've rebased the commit, so it's easier to review commit by commit since they are all independent.

X-Ryl669 commented:
Any update on the review process?

ejurgensen commented:
Reviewing isn't the most fun, but I'll try to get it done soon. I have the C stuff, @hacketiwack has the web UI.

hacketiwack commented:
I will do the review in the next few days (over the weekend). Sorry for the delay.

ejurgensen commented on src/db_init.c:
You also need to update the version number in db_init.h

X-Ryl669 replied:
Dumb me. That's why it didn't upgrade automatically for me 😭

ejurgensen commented on src/httpd_jsonapi.c:
```diff
@@ -83,6 +83,13 @@ safe_json_add_string(json_object *obj, const char *key, const char *value)
     json_object_object_add(obj, key, json_object_new_string(value));
 }
 
+static inline void
```
Remove this function, just use json_object_object_add directly like the rest of the code does

X-Ryl669 replied:
Ok

ejurgensen commented on src/httpd_jsonapi.c:
```diff
@@ -3311,6 +3336,59 @@ jsonapi_reply_library_tracks_get_byid(struct httpd_request *hreq)
+
+  ret = evbuffer_add_printf(hreq->out_body, "%s", json_object_to_json_string(reply));
+  if (ret < 0)
+    DPRINTF(E_LOG, L_WEB, "browse: Couldn't add track to response buffer.\n");
```
"browse" - looks like a copy/paste error

ejurgensen reviewed (commented):
Here's a first review covering the db and json changes. I haven't done the filescanner yet.

X-Ryl669 commented on docs/json-api.md:
````diff
@@ -1544,10 +1545,47 @@ curl -X GET "http://localhost:3689/api/library/tracks/1"
+**Endpoint**
+
+```http
+GET /api/library/lyrics/{id}
````
Thanks for reviewing!

The artwork is a specific API too, since the Track links to an Artwork URL (not the artwork itself). Lyrics are also optional (like artwork) and can be quite large. I'm not sure it's a good idea to push (optional) large data to the user, unless she actually wants it.   

In the current web implementation, it doesn't call the track endpoint currently (it's doing fine with the queue endpoint), so using the track endpoint to store the lyrics means transferring a lot of useless metadata when only the lyrics are requested.

ejurgensen replied:
Ok, but if size a concern then it might also not be great to have it as a mfi and dbmfi attribute and an extra column in the db? Then it should be made like artwork, i.e. with lyrics_url and treating the lyrics as a separate resource.

X-Ryl669 replied:
I can't say. It's an architectural decision here. Lyrics are (specially formatted) text, in the order of 1kB to 10kB for the largest one I would say. There is no processing to perform on them like for artwork (no resizing, the parsing of lyrics to synchronized sentences is done in the web client upload loading). 

So on one side, if we have to store them like artwork, it would mean extracting them and storing as LRC files or txt file or re-parsing them on the fly from the MP3 file when requested and dealing with a cache like artwork is dealt with. That's a lot of work, IMHO.

On the other side, unlike artwork because they are encoded as text, they can be stored in the DB while the files are parsed for metadata. That's a relatively small overload (typically for a 1000 song database, it increases the DB by 5MB I would say). That's less than a single song size. This saves latency because you don't have to re-launch ffmpeg to re-parse the file for metadata extraction each time it's required. 

In the implementation, I've tried to avoid changing anything to existing users, so they don't have to pay for this feature if they don't use it. By simply adding a `lyrics` key in the track JSON object **only if** lyrics were detected, it's the smallest impact on them they could have. Then, only users that have enabled the lyrics panel will fetch the /api/library/lyrics and fetch the additional data. 

If we were to move to a lyrics dynamic extraction URL like artwork, it would mean one query per track (+ one server parsing the song per track) even if it doesn't have lyrics. I think it's not interesting.

Currently, the `/api/library/lyrics` *is* the Lyrics URL, but it's fetched from cached parsing done while we had parsed the MP3 files and saved in the DB. It's not dynamic like artwork because it wouldn't make sense, we don't ask for downsized lyrics.  

I hope it's a bit clearer why I did it this way. Please let me know your opinion on this.

ejurgensen replied:
> 1kB to 10kB

That sounds quite small, so in that light I think it should just be a metadata attribute.

X-Ryl669 replied:
Ok, done in the 2nd commit, now "lyrics"'s key of the track object is filled with lyrics if present. I've updated the documentation to remove the lyric route and the web application to use track instead.

X-Ryl669 commented on src/db.c:
```diff
@@ -371,6 +372,7 @@ static const ssize_t dbmfi_cols_map[] =
     dbmfi_offsetof(channels),
     dbmfi_offsetof(usermark),
     dbmfi_offsetof(scan_kind),
+    dbmfi_offsetof(lyrics),
```
Ok, I'll do that.

X-Ryl669 commented on src/db_upgrade.c:
```diff
@@ -1227,6 +1227,26 @@ static const struct db_upgrade_query db_upgrade_v2200_queries[] =
     { U_v2200_SCVER_MINOR,    "set schema_version_minor to 00" },
   };
 
+/* ---------------------------- 28.8 -> 28.9 ------------------------------ */
```
You mean: the comment is wrong, or the `U_v2200_SCVER_MINOR` should read `U_v2201_SCVER_MINOR`?

X-Ryl669 commented:
I think it's resolved all remarks (except for design remark). I've also renamed the commit to fit the other commits name. For example files with lyrics, please follow this [link](https://transfert.free.fr/2Pv8sat) it's valid for 7 days. You'll find both synchronized, unsynchronized and tracks with no lyric.

hacketiwack commented on web-src/src/components/NavbarBottom.vue:
```diff
@@ -106,8 +106,9 @@
-    <div class="navbar-brand is-flex-grow-1">
-      <navbar-item-link :to="{ name: 'queue' }" exact class="mr-auto">
+   <div class="navbar-brand is-flex-grow-1">
+      <group class="navbar-item is-expanded is-justify-content-left is-no-basis">
```
Where does this group element come from?
It doesn't appear to be an HTML element.
Could you please explain?
Thanks

X-Ryl669 replied:
The bottom bar is functionally 3 groups of items. There's one (group of) item on the left for switching between queue mode and "now playing mode", the center group of playing buttons and the right group for toggling lyrics and showing the output panes.

Initially, there was an hack by using bulma's navbar (that was not made for splitting the bar in 3 groups) done by using right margin auto on the left item, and left margin on the right item. This worked because the number of elements was odd and all the same size (except for the center).

The change I've made is to switch to bulma's flex classes, when each group of items must be a direct children of the navbar-brand. By using flex, the centering of items in the box is now clearly defined (justify left, center, and right) and doesn't depend on the item size and count. 

I've had to add `is-no-basis` since it's bulma doesn't have a class for `flex-basis:0`. 

I've used the element `group` because it's a group. You are allowed to use any tag name in HTML5, and I'm not using any CSS rule with that tag name. It's possible to use a `div` if you prefer.

hacketiwack replied:
Thanks for the explanation.
I thought the group was some sort of special element I didn't know.
For clarity, and because it's not a specific component like the element `<mdicon>`, I would recommend using a simple `div` in this situation.

X-Ryl669 replied:
Done

hacketiwack commented on web-src/src/components/NavbarBottom.vue:
```diff
@@ -271,6 +282,7 @@ import PlayerButtonConsume from '@/components/PlayerButtonConsume.vue'
 import PlayerButtonNext from '@/components/PlayerButtonNext.vue'
 import PlayerButtonPlayPause from '@/components/PlayerButtonPlayPause.vue'
 import PlayerButtonPrevious from '@/components/PlayerButtonPrevious.vue'
+import PlayerButtonLyrics from '@/components/PlayerButtonLyrics.vue'
```
Can you please order the imports alphabetically?
Thanks in advance.

hacketiwack commented on web-src/src/components/NavbarBottom.vue:
```diff
@@ -287,6 +299,7 @@ export default {
     PlayerButtonNext,
     PlayerButtonPlayPause,
     PlayerButtonPrevious,
+    PlayerButtonLyrics,
```
Can you please order the components alphabetically?
Thanks in advance.

X-Ryl669 replied:
Done

hacketiwack commented on web-src/src/mystyles.scss:
```diff
@@ -3,6 +3,29 @@
+.navbar.is-dark .navbar-brand .navbar-item, 
+.navbar.is-white.is-dark .navbar-brand .navbar-item, 
+{
+  color: #fff;
```
Can you please use the variables for colours?
Thanks.

hacketiwack commented on web-src/src/mystyles.scss:
```diff
@@ -3,6 +3,29 @@
+
+/* Fix bulma issue for a group under a navbar-brand container (contrary to documentation) */
+.navbar.is-white .navbar-brand > .navbar-item *:focus,  .navbar.is-white .navbar-brand >  .navbar-item *:hover, .navbar.is-white .navbar-brand > .navbar-item *.is-active {
+  color: #0a0a0a;
```
Can you please use the variables for colours?
Thanks.

X-Ryl669 replied:
Which one ?

This is a copy and paste from bulma's code to fix a bug in their declaration. It should work with a ["container with anything you want"](https://bulma.io/documentation/components/navbar/#navbar-item) but it doesn't in reality. 

I've fixed the CSS rule to support a container correctly. The content of the rule is a copy and paste from bulma's CSS. 

I haven't tried using $white or $grey because I don't know if this variable that's defined in bulma is available in this file. If it is, which one should I use ?

hacketiwack replied:
I would use `$white` for the white (#ffffff) and `$black` for the kind of "dark grey" (#0a0a0a).
The Bulma variables can be used at this place, because they're imported with the statement `@import 'bulma/bulma.sass';`

I also made a slight change in the use of color variables in the main branch, because they were referenced incorrectly: `#{white}` instead of `$white` and `#{$danger}` instead of `$danger`

The Bulma colors are available [here](https://bulma.io/documentation/overview/colors/).

X-Ryl669 replied:
Done

hacketiwack commented on web-src/src/mystyles.scss:
```diff
@@ -3,6 +3,29 @@
+.navbar.is-dark .navbar-brand > .navbar-item:focus, .navbar.is-dark .navbar-brand > .navbar-item:hover, .navbar.is-dark .navbar-brand >  .navbar-item.is-active 
+.navbar.is-dark .navbar-brand > .navbar-item *:focus, .navbar.is-dark .navbar-brand > .navbar-item *:hover, .navbar.is-dark .navbar-brand >  .navbar-item *.is-active  {
+  background-color: #292929;
+  color: #fff;
```
Can you please use the variables for colours?
Thanks.

hacketiwack reviewed (changes requested):
Thank you for taking into account the comments and make the according changes.

hacketiwack commented on web-src/src/components/Lyrics.vue:
```diff
@@ -0,0 +1,188 @@
+    return {
+      scroll: {},
+      // lineHeight: 42,
+      scrollAbled: true, // stop scroll to element when touch
```
Do you mean scroll enabled with this word?

X-Ryl669 replied:
It'd `able to scroll` or `canAutoScroll`.

This is used to disable auto scrolling when the user scrolls the lyrics manually for a short period of time.  Without this, a synchronized lyric will scroll by itself and disturb a user wanting to read them by her own.

hacketiwack replied:
Shouldn't this be variable simply named `autoScroll`?

X-Ryl669 replied:
Sure, I will rename it.

X-Ryl669 replied:
Done

hacketiwack commented on web-src/src/components/NavbarBottom.vue:
```diff
@@ -153,15 +156,23 @@
       />
+      </group>
+      <group class="navbar-item is-expanded is-justify-content-right is-no-basis">
+      <player-button-lyrics
```
Now that I'm looking at the interface, I'm wondering if this button should be placed here. It could be ok for now, but definitely the bottom bar should be redesigned.

I know that there are currently some impracticalities with the volume control as mentioned by @ejurgensen in this [work item](https://github.com/owntone/owntone-server/issues/1442).

I would suggest to mostly copy the design of Spotify or Jellyfin.

X-Ryl669 replied:
But this is mostly how [spotify is doing it](https://beebom.com/wp-content/uploads/2021/11/mic-icon-spotify-lyrics.jpg?quality=75&strip=all). The lyrics button is on the bottom right. Jellyfin doesn't seem to support lyrics for music (or I haven't found how on the demo).

Amazon music "hides" the lyrics on the album thumbnails on the bottom left (you click the thumbnail and the lyrics pane appears to replace the complete interface). I don't like this, you can't detect this feature unless someone told you so. When the lyrics are showing, you don't have access to the interface, which is also bad, IMHO. Lyrics shouldn't hide the player information, or you'll spend time toggling the lyrics on & off for each track.

I agree that the (global) volume bar is missing on the bottom bar, and hopefully, my changes to the flex layout now allows to add such widget without breaking the centering of the player's icons. 

So it should be possible to add this after this is merged.

hacketiwack replied:
Yes, I agree with the points above.
It was more of a general comment for the usability of OwnTone web interface that we'll have to improve in the future.
Let's go ahead with your proposal.

X-Ryl669 replied:
Done

X-Ryl669 commented:
Ok, I've reworked the commits with all your requested change (API + Web changes). 

@hacketiwack I've squashed your commits in the last one so it's semantically only one with the "convention" naming for the commit message. Feel free to check if I haven't miss anything. Thank you!

hacketiwack commented:
All good for me 👍🏼

ejurgensen commented on src/library/filescanner_ffmpeg.c:
```diff
@@ -295,44 +310,64 @@ extract_metadata_core(struct media_file_info *mfi, AVDictionary *md, const struc
+          if ((mdt->value == NULL) || (strlen(mdt->value) == 0))
+            break;
+
+          if (memcmp(mdt->key, "lyrics-", sizeof("lyrics-") - 1) == 0)
```
Move to match_metadata and use strcmp/strncmp for strings

X-Ryl669 replied:
No, here it's a partial match, we are searching for the "lyrics-" prefix in a larger key (like "lyrics-XXX" or "lyrics-fra"). match_metadata search for a perfect match. strncmp might be suitable.

ejurgensen commented on src/library/filescanner_ffmpeg.c:
```diff
@@ -295,44 +310,64 @@ extract_metadata_core(struct media_file_info *mfi, AVDictionary *md, const struc
+  mdcount = 0;    
+  
+  /* Cache this search once to avoid doing it per metadata key */
+  lyricsindex = match_metadata("lyrics", md_map);
```
If I understand this correctly it just saves cycles in the case where a file has lyrics in many different languages? I'm not sure that's worth optimizing. Multiple languages will cause another issue, which is that we aren't necessarily picking the one the user wants. Basically I think that use case is so exotic that we shouldn't cater for it.

X-Ryl669 replied:
No: this finds the index of Owntone's "lyrics" key in the static md_map (which is static and fixed at compile time). So it's done once (currently per file, could be moved out to be done per run) instead of being done once per metadata found in the file.

The idea is:
```python
   for (each metadata found in file):
      if metadataKey starts with "lyrics":
         lyricsIndex = find the index in md_map for the key "lyrics"   << This is static and knowable at compile time
         save the lyrics in the metadata array at index lyricsIndex
     else:
        index = find the index in md_map for the metadataKey
        save the metadata in the metadata array at index index 
```

Here, the lyricsIndex is known at compile time, so there's no point in searching it for each lyrics (and logically even for any file). Ideally, it could even be saved once for all when starting the whole parsing process, but I haven't done that yet.

X-Ryl669 replied:
Moved to a static index so it's only done once per runtime.

ejurgensen replied:
Maybe I didn't explain myself well enough. I did understand that it saved the fixed index. Here's examples to show you what I mean:

- 5 metadata, none of them lyrics: 6 iterations (one to find the fixed index of lyrics + 5 x match_metadata)
- 5 metadata , one of them lyrics:  5 iterations (one to find the fixed index of lyrics + 4 x match_metadata)
- 5 metadata , two of them lyrics:  5 iterations (one to find the fixed index of lyrics + 3 x match_metadata)

If you instead just set the key to "lyrics" for any variation of "lyrics-" and then proceed as normal, all the above will require 5 iterations and the code will be more simple.

X-Ryl669 replied:
Ok. That's right. I've changed it so now it's done once for all, so it'll only cost one search per launch of the program and none afterward. Should be better now.

ejurgensen commented on src/library/filescanner_ffmpeg.c:
```diff
@@ -295,44 +310,64 @@ extract_metadata_core(struct media_file_info *mfi, AVDictionary *md, const struc
+          if (memcmp(mdt->key, "lyrics-", sizeof("lyrics-") - 1) == 0)
+            i = lyricsindex; /* Replace by the generic lyrics metadata map */
+          else
+            /* Check if matches one of the mapped key */
```
Remove comment, doesn't add anything

ejurgensen commented on src/library/filescanner_ffmpeg.c:
```diff
@@ -295,44 +310,64 @@ extract_metadata_core(struct media_file_info *mfi, AVDictionary *md, const struc
+        }
+      while(0); /* Used so that 'break' works inside the scope */
+
+      mdt = av_dict_get(md, "", mdt, AV_DICT_IGNORE_SUFFIX);
```
av_dict_iterate()

X-Ryl669 replied:
Ok made the change. However, got a warning for the CI since av_dict_iterate is only on recent ffmpeg:
```bash
library/filescanner_ffmpeg.c:364:9: warning: implicit declaration of function ‘av_dict_iterate’; did you mean ‘av_codec_iterate’? [-Wimplicit-function-declaration]
  364 |   mdt = av_dict_iterate(md, NULL);
      |         ^~~~~~~~~~~~~~~
      |         av_codec_iterate
library/filescanner_ffmpeg.c:364:7: warning: assignment to ‘const AVDictionaryEntry *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  364 |   mdt = av_dict_iterate(md, NULL);
      |       ^
library/filescanner_ffmpeg.c:368:11: warning: assignment to ‘const AVDictionaryEntry *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  368 |       mdt = av_dict_iterate(md, mdt);
```

Works for me on Arch

ejurgensen replied:
Ok, then we have to do a configure check for av_dict_iterate() and if it isn't present, define it from av_dict_get. I can add this.

X-Ryl669 replied:
av_dict_iterate is just a wrapper over av_dict_get, so I'll change it back to av_dict_get, it should work everywhere without a configure check.

ejurgensen commented on src/httpd_jsonapi.c:
```diff
@@ -3311,6 +3313,7 @@ jsonapi_reply_library_tracks_get_byid(struct httpd_request *hreq)
   return HTTP_OK;
 }
 
+
```
Remove newline

X-Ryl669 commented on src/library/filescanner_ffmpeg.c:
```diff
@@ -295,44 +310,64 @@ extract_metadata_core(struct media_file_info *mfi, AVDictionary *md, const struc
-	    {
-	      ret = safe_atou32(mdt->value, intval);
-	      if (ret < 0)
-		continue;
```
I'm pretty sure it's a bug here. If it failed converting to int, shouldn't mdcount be decremented ?

ejurgensen replied:
Yes, just looks like the mdcount++ should have been placed lower (it just seems to be used for logging, however).

X-Ryl669 replied:
Ok, will change this tomorrow so it doesn't return 1 but 0 in that case.

X-Ryl669 commented:
I have applied all changes requested. Please review for the potential bug I've mentioned, if it's correct, I'll remove the comment.

X-Ryl669 commented:
I'm sorry I haven't understood your last message. What change did I miss? How can I help you better?

ejurgensen commented:
Merged it now, thanks for the contribution. As you can see, the parts I was missing was using av_dict_iterate and not doing the pre-search for the lyrics index. I also fixed up a few other things (e.g., there was a potential bug in the old code where the value from ffmpeg dictionary was being modified, but the docs say not to do that).

Hope it works, I have to admit I didn't actually test it with a lyrics file.

X-Ryl669 commented:
Testing right now. Please wait.

ejurgensen commented:
I notice now that the lyrics icon is showing for all tracks, even if they don't have lyrics. Can you change it so it only shows if there are lyrics?

X-Ryl669 commented:
It's working for me. 

Just a remark: the systemd service file isn't working well for me. It should include a `User=` and `Group=` rule, since without this, it'll run as root and then setuid to whatever user you've set up in the `owntone.conf`. Yet the group is then the (low privilege) user that doesn't have read access to the song library. 

If I understand POSIX user management correctly, the songs in the library should be owned by a "music" group whose user are allowed to read and/or write to them. I want to be able to set up this user in either the conf file (or in the service file).

X-Ryl669 commented:
> Can you change it so it only shows if there are lyrics?

This starts an usability issue: If you use the lyric mode (by clicking on the button) and song A has lyrics, then song B hasn't and then song C has too. You don't want to disable the lyrics mode when reaching song B because then you'd have to enable it again in song C. 

The other player, like spotify, amazon music, works the same. When you have toggled the lyrics, it stays on even for tracks without lyrics. 

What do you think, should we act differently?

ejurgensen commented:
The current solution is not an option in view. I personally would prefer to have as few buttons on the interface as possible to keep it clean. Would there be a fifth option to not have any button, but just auto-display the lyrics if they are available? With a "close" button if they user doesn't want to see them.

@hacketiwack what do you think?

ejurgensen commented:
I've reverted the UI changes, in addition to the button issue I noticed that the linter was had some issues with them:

```
/home/runner/work/owntone-ghatest/owntone-ghatest/web-src/src/components/Lyrics.vue
  11:7  error  Elements in iteration expect to have 'v-bind:key' directives  vue/require-v-for-key
  23:9  error  Component name "Lyrics" should always be multi-word           vue/multi-word-component-names
  57:9  error  Unexpected side effect in "lyricIndex" computed property      vue/no-side-effects-in-computed-properties
  58:9  error  Unexpected side effect in "lyricIndex" computed property      vue/no-side-effects-in-computed-properties

/home/runner/work/owntone-ghatest/owntone-ghatest/web-src/src/components/PlayerButtonLyrics.vue
  19:8  error  'webapi' is defined but never used  no-unused-vars
```

Please fix and submit a new PR

hacketiwack commented:
I went through the last version of the source code for the web part. Two main points:

1. It still needs to be prettified with the command `npm run format`.
2. The bottom navigation bar is not working properly:
    - On mobile devices, the title of the track is not truncated anymore when the size is too small (see below).
        ![Truncated titles](https://github.com/owntone/owntone-server/assets/13329768/6c9eedf9-5247-4130-a6c5-c73c1e6d8a09)       
    - The buttons are now too large (see below).
![Too large buttons](https://github.com/owntone/owntone-server/assets/13329768/561645fe-945e-4410-951e-9e0e45dd4aab)

For point 2, I highly recommend to keep the old navigation bar and add the lyrics button in the menu where there is the volume. We will improve the navigation bar in another release.
This is how it should look like:
![Proposal](https://github.com/owntone/owntone-server/assets/13329768/ab5932a5-3ebc-4ec2-8e58-33553107256c)
